### DISCUSSION TITLE
Does "data size" in VertiPaq Analyzer show compressed data size?

### DISCUSSION THREAD

**Discussion by neednlab** [DaxStudio/DaxStudio#541] [Q&A]
Hello,

I'm new to DAX studio and Power BI. I want to test the compression ratio of data stored in VertiPaq(Power BI). 
Firstly I created 2 tables with the same schema in SQL Server, wide_fact1 and wide_fact2. 
Then insert exact the same 10Ms rows into both except the first column( a clustered index).  
Data in wide_table1 is sorted randomly while the same value is stored continuously in wide_table2 as following.
![table1_vs_table2](https://user-images.githubusercontent.com/41602171/107019719-fa9d2200-67dc-11eb-8b84-ff2a98331a66.png)
Finally I've extracted data in both tables with "import mode" in Power BI desktop. As I expected, wide_fact1 pbix file stored 223MB and wide_fact2 pbix file is 115MB.
But I'm confused that when I opened VertiPaq Analyzer in DAX studio to analyze the data size, there is almost no difference between the two.
![vertipaq_wide_fact1](https://user-images.githubusercontent.com/41602171/107021869-a2b3ea80-67df-11eb-855b-3c3e802805b3.png)
![vertipaq_wide_fact2](https://user-images.githubusercontent.com/41602171/107021897-aa738f00-67df-11eb-909a-1dfd5cd7a86f.png)
Is there something wrong? 

Thanks
Needn

**Comment by marcosqlbi** [comment]
VertiPaq sort rows in memory before compression. The difference you might see could be caused by a different compression of the PBIX (which compresses data that is not compressed by VertiPaq), even though the delta is too big to be justified by such a difference.
I suggest you double-check you didn't include other data in the largest PBIX file.

**Reply by neednlab** [reply]
Hi @marcosqlbi Thank you for your quick reply.
I'm sure that there is only one table in both PBIX files, you can see "% DB" shows 99.98%(there is another auto created date table) and 100%.

Here is my DLL script (SQL Server)
    DECLARE @n AS BIGINT = 1000000;

    WITH
    L0   AS(SELECT 1 AS c UNION ALL SELECT 1),
    L1   AS(SELECT 1 AS c FROM L0 AS A CROSS JOIN L0 AS B),
    L2   AS(SELECT 1 AS c FROM L1 AS A CROSS JOIN L1 AS B),
    L3   AS(SELECT 1 AS c FROM L2 AS A CROSS JOIN L2 AS B),
    L4   AS(SELECT 1 AS c FROM L3 AS A CROSS JOIN L3 AS B),
    L5   AS(SELECT 1 AS c FROM L4 AS A CROSS JOIN L4 AS B),
    Nums AS(SELECT ROW_NUMBER() OVER(ORDER BY (SELECT 0)) AS n FROM L5)
    SELECT n INTO dbo.Nums
    FROM Nums WHERE n <= @n;


    CREATE TABLE SS_FACT
    (
        SS_ID BIGINT NOT NULL PRIMARY KEY,
        A_ID INT NOT NULL,
        B_ID INT NOT NULL,
        C_ID INT NOT NULL
    );

    CREATE TABLE WIDE_FACT1
    (
        SS_ID BIGINT NOT NULL PRIMARY KEY,
        A_ID INT NOT NULL,
        B_ID INT NOT NULL,
        C_ID INT NOT NULL,
        A_1	CHAR(36),
        A_2	CHAR(36),
        A_3	CHAR(36),
        A_4	CHAR(36),
        A_5	CHAR(36),
        B_1	CHAR(36),
        B_2	CHAR(36),
        B_3	CHAR(36),
        B_4	CHAR(36),
        B_5	CHAR(36),
        C_1	CHAR(36),
        C_2	CHAR(36),
        C_3	CHAR(36),
        C_4	CHAR(36),
        C_5	CHAR(36)
    );

    CREATE TABLE WIDE_FACT2
    (
        SS_ID BIGINT NOT NULL PRIMARY KEY,
        A_ID INT NOT NULL,
        B_ID INT NOT NULL,
        C_ID INT NOT NULL,
        A_1	CHAR(36),
        A_2	CHAR(36),
        A_3	CHAR(36),
        A_4	CHAR(36),
        A_5	CHAR(36),
        B_1	CHAR(36),
        B_2	CHAR(36),
        B_3	CHAR(36),
        B_4	CHAR(36),
        B_5	CHAR(36),
        C_1	CHAR(36),
        C_2	CHAR(36),
        C_3	CHAR(36),
        C_4	CHAR(36),
        C_5	CHAR(36)
    );


    CREATE TABLE SS_DIM_A
    (
        A_ID INT NOT NULL,
        A_1	CHAR(36),
        A_2	CHAR(36),
        A_3	CHAR(36),
        A_4	CHAR(36),
        A_5	CHAR(36)

    );

    CREATE TABLE SS_DIM_B
    (
        B_ID INT NOT NULL,
        B_1	CHAR(36),
        B_2	CHAR(36),
        B_3	CHAR(36),
        B_4	CHAR(36),
        B_5	CHAR(36)

    );

    CREATE TABLE SS_DIM_C
    (
        C_ID INT NOT NULL,
        C_1	CHAR(36),
        C_2	CHAR(36),
        C_3	CHAR(36),
        C_4	CHAR(36),
        C_5	CHAR(36)

    );

Then I populated data using the following script

    TRUNCATE TABLE SS_DIM_A;
    INSERT INTO SS_DIM_A
    SELECT 
        N,NEWID(),NEWID(),NEWID(),NEWID(),NEWID()
    FROM Nums
    WHERE N BETWEEN 1 AND 10;
    GO

    TRUNCATE TABLE SS_DIM_B;
    INSERT INTO SS_DIM_B
    SELECT 
        N,NEWID(),NEWID(),NEWID(),NEWID(),NEWID()
    FROM Nums
    WHERE N BETWEEN 1 AND 100;

    TRUNCATE TABLE SS_DIM_C;
    INSERT INTO SS_DIM_C
    SELECT 
        N,NEWID(),NEWID(),NEWID(),NEWID(),NEWID()
    FROM Nums
    WHERE N BETWEEN 1 AND 10000;

    TRUNCATE TABLE [dbo].[SS_FACT]
    INSERT INTO [dbo].[SS_FACT] WITH(TABLOCKX)
    SELECT
        ROW_NUMBER() OVER(ORDER BY (SELECT 1)) AS ROW_ID,
        A.A_ID,
        B.B_ID,
        C.C_ID
    FROM 
    [dbo].[SS_DIM_A] AS A
    CROSS JOIN
    [dbo].[SS_DIM_B] AS B
    CROSS JOIN
    [dbo].[SS_DIM_C] AS C;
    GO

    TRUNCATE TABLE [dbo].[WIDE_FACT1];
    SELECT
        A.A_ID,
        B.B_ID,
        C.C_ID,
        A.A_1,
        A.A_2,
        A.A_3,
        A.A_4,
        A.A_5,
        B.B_1,
        B.B_2,
        B.B_3,
        B.B_4,
        B.B_5,
        C.C_1,
        C.C_2,
        C.C_3,
        C.C_4,
        C.C_5,
        CONVERT(CHAR(36),NEWID()) AS NEW_ID
    INTO #T
    FROM 
    [dbo].[SS_DIM_A] AS A
    CROSS JOIN
    [dbo].[SS_DIM_B] AS B
    CROSS JOIN
    [dbo].[SS_DIM_C] AS C;

    INSERT INTO [dbo].[WIDE_FACT1] WITH(TABLOCKX)
    SELECT 
        ROW_NUMBER() OVER(ORDER BY NEW_ID) AS ROW_ID
        ,[A_ID]
        ,[B_ID]
        ,[C_ID]
        ,[A_1]
        ,[A_2]
        ,[A_3]
        ,[A_4]
        ,[A_5]
        ,[B_1]
        ,[B_2]
        ,[B_3]
        ,[B_4]
        ,[B_5]
        ,[C_1]
        ,[C_2]
        ,[C_3]
        ,[C_4]
        ,[C_5]
    FROM #T;
    GO

    TRUNCATE TABLE [dbo].[WIDE_FACT2];
    INSERT INTO [dbo].[WIDE_FACT2] WITH(TABLOCKX)
    SELECT
        ROW_NUMBER() OVER(ORDER BY (SELECT 1)) AS ROW_ID,
        A.A_ID,
        B.B_ID,
        C.C_ID,
        A.A_1,
        A.A_2,
        A.A_3,
        A.A_4,
        A.A_5,
        B.B_1,
        B.B_2,
        B.B_3,
        B.B_4,
        B.B_5,
        C.C_1,
        C.C_2,
        C.C_3,
        C.C_4,
        C.C_5
    FROM 
    [dbo].[SS_DIM_A] AS A
    CROSS JOIN
    [dbo].[SS_DIM_B] AS B
    CROSS JOIN
    [dbo].[SS_DIM_C] AS C

Do you think it's normal for such a big difference?
Thanks
Needn

**Reply by marcosqlbi** [reply]
I'm surprised by the differences in PBIX size. Can you rename the PBIX files to ZIP and analyze where the difference is? Probably in another file corresponding to the AS backup.
Also test VertiPaq Analyzer after you opened the PBIX file, not after a refresh.

**Reply by neednlab** [reply]
I tried to unzip the PBIX file by 7-zip, then found DataModel file is 223MB & 115MB. I have no idea how to investigate further 
Can you please download those 2 files and take a look？ Thanks.

https://neednlab.blob.core.windows.net/public/wide_table1.pbix
https://neednlab.blob.core.windows.net/public/wide_table2.pbix

**Reply by marcosqlbi** [reply]
It seems the larger PBIX file has additional data that is not needed, but I don't have a good explanation.
I would suggest you try creating the wide_table1 from scratch to see if the final PBIX file is always the same - but we don't have tools to investigate further.

**Reply by neednlab** [reply]
Hi @marcosqlbi , I've tested with wide_table3 (the same as wide_table1 )and wide_table4(a columnstore index), and got the same result as before.
The same total size in the VertiPaq Analyzer, but different in PBIX files on the disk.
It looks weird, I will post it to community later.
![wide_table1](https://user-images.githubusercontent.com/41602171/107107015-88265380-6869-11eb-99b9-6f9a06bfaaa8.png)
![wide_table2](https://user-images.githubusercontent.com/41602171/107107019-8a88ad80-6869-11eb-869c-e611b196dbde.png)
![pbix_size](https://user-images.githubusercontent.com/41602171/107107021-8ceb0780-6869-11eb-8c65-24f2ea560dac.png)


Actually, I‘m trying to reduce the data size of Power BI dataset, I've suggested our data analyst use the star schema instead of a single wide table. But I need some evidence the star schema is useful, that's why I doing this test.
I'm sure that the star schema has so many benefits not only the data reduction. however it will be affected by the high data compression ratio. Am I right?

By the way, I'm reading the book "Analyzing Data With Microsoft Power BI and Power Pivot for Excel". I've noticed that you are the author! that's amazing! It is really very helpful for learning Power BI data modeling. Thanks so much.
